Only deploy container engine if needed on etcd nodes

VannTen opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md and developer guide https://git.k8s.io/community/contributors/devel/development.md
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/guide/release-notes.md
6. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What type of PR is this?**
/kind bug

**What this PR does / why we need it**:

If the etcd cluster is separate and the etcd_deployment_type is "host",
there is no need for a container engine on the etcd nodes


**Which issue(s) this PR fixes**:

Fixes #7314
The issue was closed by its author, but it is not solved

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".
-->
```release-note
Container engine is no longer installed on separate etcd nodes when using
`etcd_deployment_type: host`
```

champtar commented on cluster.yml:
```diff
@@ -38,7 +38,11 @@
+    - role: "container-engine"
+      tags: "container-engine"
+      when:
+        - deploy_container_engine|default(true)
+        - etcd_deployment_type != "host" or "k8s-cluster" in group_names
```
I think you should define `deploy_container_engine` in `roles/kubespray-defaults/defaults/main.yaml`, because `deploy_container_engine` is used at multiple locations
something like
```
deploy_container_engine: etcd_deployment_type != "host" or inventory_hostname in groups['kube-cluster']
```

VannTen commented:
On Thu, Apr 22, 2021 at 06:53:51AM -0700, Etienne Champetier wrote:
> @champtar requested changes on this pull request.
> 
> 
> 
> > +        - deploy_container_engine|default(true)
> +        - etcd_deployment_type != "host" or "k8s-cluster" in group_names
> 
> I think you should define `deploy_container_engine` in `roles/kubespray-defaults/defaults/main.yaml`, because `deploy_container_engine` is used at multiple locations
> something like
> ```
> deploy_container_engine: etcd_deployment_type != "host" or inventory_hostname in groups['kube-cluster']
> ```

With the current state (of this PR), it's impossible to override it for
external etcd nodes using the host method.

If I'm correct, using the default would make it possible. (Correct me
if I'm mistaken).
Should it be ? (possible, I mean). I don't see a case where kubespray
should do it, but I may have missed some use cases.
I'll defer to your opinion on that.


By the way, regardless of the path which will be chosen, I realized I'll
need to update the other location where the role is imported with that
check.


-- 
Max Gautier
Consultant Cloud
***@***.***
Objectif Libre

champtar commented:
My point is don't update all the locations at all and just define `deploy_container_engine` in `roles/kubespray-defaults/defaults/main.yaml`
Also you need to sign the CLA ;)

VannTen commented:
I modified the pr according to this.
I updated the (not that many) location which were using the 'default(true)',
since it does not make sense to have it if we define the default in
kubespray-defaults.

I got your point, but mine was that the behavior is slighly different.

Regarding the CLA, I need a CCLA (I'm doing this on company time), and I've
asked my management if we already have one or need to do the sign up.

champtar commented on roles/kubespray-defaults/defaults/main.yaml:
```diff
@@ -230,6 +230,9 @@ kube_api_aggregator_routing: false
 kube_profiling: false
 
+# Whether to deploy the container engine
+deploy_container_engine: inventory_hostname in groups['k8s-cluster'] or etcd_deployment_type != 'host'
```
now that I see it I have a doubt ...
```suggestion
deploy_container_engine: {{ inventory_hostname in groups['k8s-cluster'] or etcd_deployment_type != 'host' }}
```

oomichi replied:
Is it fine to skip this @champtar point?

Anyways we need to replace `k8s-cluster` with `k8s_cluster` because of the commit https://github.com/kubernetes-sigs/kubespray/commit/360aff4a571f4a8ac7603a4a68223e744fad1209

rtsp replied:
I have a problem while upgrading the etcd in #8381 so I trying to debug by printing out the value of `deploy_container_engine`

```yaml
ansible.builtin.debug:
  msg: "deploy_container_engine={{ deploy_container_engine }} host={{ inventory_hostname }}, etcd={{ etcd_deployment_type }}
```

```
TASK [deploy_container_engine]
ok: [callisto-e3] => {
  "msg": "deploy_container_engine=inventory_hostname in groups['k8s_cluster'] or etcd_deployment_type != 'host' host=callisto-e3, etcd=host"
}
```

It's turn out that the variable is literally a string from `kubespray/roles/kubespray-defaults/defaults/main.yaml` so it evaluated as `True` in `{ role: container-engine, tags: "container-engine", when: deploy_container_engine }` and cause the container-engine role to run 


After I apply a fix in #8386 the value is now boolean as expected and the problem solved.

```
TASK [deploy_container_engine] 
ok: [callisto-e1] => {
  "msg": "deploy_container_engine=False host=callisto-e1, etcd=host"
}
```

VannTen commented:
> now that I see it I have a doubt ...
> ```suggestion
> deploy_container_engine: {{ inventory_hostname in groups['k8s-cluster'] or etcd_deployment_type != 'host' }}
> ```
Well, that would need quotes around the brackets. But I don't think we
need to interpolate here ?

champtar commented:
@VannTen any progress on the CLA signing ?

champtar commented:
That answers the CLA signing question, still need rebase

VannTen commented:
I'll do that.

VannTen commented:
Is there anything missing from my side to proceed ?

floryut commented:
> Is there anything missing from my side to proceed ?

I've restarted tests as there was a temp issue with the CI on your PR

floryut commented:
/cc @champtar
All good for you ?

floryut commented:
/cc @champtar @oomichi

VannTen commented:
I've updated the PR to use the new naming scheme. The failing job does not seem related to me (certificate problem). Is there anything missing from my side ?
